cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,446 @@
+     * @param ch_out The output channel to send messages to.
+     * @return A coroutine representing the task.
+     */
+    Node send_task(Context& ctx, size_t& self_next_idx, std::shared_ptr<Channel> ch_out) {
+        ShutdownAtExit ch_shutdown{ch_out};
+        SetChannelIdxInvalidAtExit set_ch_idx_invalid{
+            .fanout = this, .self_next_idx = self_next_idx
+        };
+        co_await ctx.executor()->schedule();
+
+        size_t n_available_messages = 0;  // number of messages available to send
+        std::vector<std::reference_wrapper<Message>> messages_to_send;
+        while (true) {
+            {
+                auto lock = co_await mtx.scoped_lock();
+                co_await data_ready.wait(lock, [&] {
+                    // irrespective of no_more_input, update the end_idx to the total
+                    // number of messages
+                    n_available_messages = recv_messages.size();
+                    return no_more_input || self_next_idx < n_available_messages;
+                });
+                if (no_more_input && self_next_idx == n_available_messages) {
+                    // no more messages will be received, and all messages have been sent
+                    break;
+                }
+                // stash msg references under the lock
+                messages_to_send.reserve(n_available_messages - self_next_idx);
+                for (size_t i = self_next_idx; i < n_available_messages; i++) {
+                    messages_to_send.emplace_back(recv_messages[i]);
+                }
+            }
+
+            for (auto const& msg : messages_to_send) {
+                RAPIDSMPF_EXPECTS(!msg.get().empty(), "message cannot be empty");
+
+                auto res = ctx.br()->reserve_or_fail(
+                    msg.get().copy_cost(), try_memory_types(msg.get())
+                );
+                if (!co_await ch_out->send(msg.get().copy(res))) {
+                    // Failed to send message. Could be that the channel is shut down.
+                    // So we need to abort the send task, and notify the process input
+                    // task
+                    co_await set_ch_idx_invalid.set_channel_idx_invalid();
+                    co_return;
+                }
+            }
+            messages_to_send.clear();
+
+            // now next_idx can be updated to end_idx, and if !no_more_input, we need to
+            // request the recv task for more data
+            auto lock = co_await mtx.scoped_lock();
+            self_next_idx = n_available_messages;
+            if (self_next_idx == recv_messages.size()) {
+                if (no_more_input) {
+                    // no more messages will be received, and all messages have been sent
+                    break;
+                } else {
+                    // request more data from the input channel
+                    lock.unlock();
+                    co_await request_data.notify_one();
```

Comment:
I was reminded what was bugging me here. If there are multiple tasks running `notify_one` (or `notify_all`) on the same condvar, we can have lost wakeups unfortunately. I have tried to fix this in libcoro here: https://github.com/jbaldwin/libcoro/pull/416

Reply:
I think the reason for that is that these tasks are typically not all firing at once. I think we are probably OK to merge now, but keep that in mind in case we see issues.